### PR TITLE
up_down_mixer: Convert module to use the module interface

### PR DESCRIPTION
**Pull request by ranj063**
Use the module_adapter interface instead of comp_drv ops.

Signed-off-by: Ranjani Sridharan <ranjani.sridharan@linux.intel.com>

### CODE REVIEW

File: src/audio/up_down_mixer/up_down_mixer.c [hunk 1]
```diff
@@ -320,232 +320,136 @@ static int init_mix(struct comp_dev *dev,
...
 	if (ret < 0) {
-		comp_free(dev);
-		return NULL;
+		comp_err(dev, "up_down_mixer_init(): failed to initialize up_down_mix");
+		goto err;
 	}
-	dev->state = COMP_STATE_READY;
 
-	return dev;
+	return ret;
```
Comment (lyakh) [inline src/audio/up_down_mixer/up_down_mixer.c:393]: this can be just `return 0`

**Review by lgirdwood — COMMENTED** [review]
@lyakh good for sparse ?
@kv2019i any comments ?

File: src/audio/up_down_mixer/up_down_mixer.c [hunk 2]
```diff
@@ -320,232 +320,136 @@ static int init_mix(struct comp_dev *dev,
...
+	comp_dbg(dev, "up_down_mixer_process()");
 
-	mix_frames = MIN(audio_stream_get_avail_frames(&source_c->stream),
-			 audio_stream_get_free_frames(&sink_c->stream));
+	mix_frames = mod->input_buffers[0].size;
 
-	source_bytes = mix_frames * audio_stream_frame_bytes(&source_c->stream);
-	sink_bytes = mix_frames * audio_stream_frame_bytes(&sink_c->stream);
+	source_bytes = mix_frames * audio_stream_frame_bytes(mod->input_buffers[0].data);
+	sink_bytes = mix_frames * audio_stream_frame_bytes(mod->output_buffers[0].data);
```
Comment (lyakh) [inline src/audio/up_down_mixer/up_down_mixer.c:426]: aren't `input_buffer` / `output_buffer` function parameters already pointing to these arrays?
```
	ret = module_process(mod, mod->input_buffers, num_input_buffers,
			     mod->output_buffers, mod->num_output_buffers);
```

**Comment by lyakh** [comment]
> @lyakh good for sparse ?

@lgirdwood don't see any address-space violations, sparse output also doesn't contain any new failures

**Comment by lgirdwood** [comment]
@ranj063 can you check CI. Thanks

File: src/audio/up_down_mixer/up_down_mixer.c (hunk 2 above)
Comment (kv2019i) [reply src/audio/up_down_mixer/up_down_mixer.c:426]: @ranj063 ? This seems a generic issue? Why pass input and output buffers to process() if they are already set up to "struct processing_module" in the prepare step. It seems this is a module adapter specific thing and modules themselves should use the input/outputs passed to process instead.

**Review by kv2019i — COMMENTED** [review]
Seems ok, issue @lyakh 's raised is interesting.

File: src/audio/up_down_mixer/up_down_mixer.c [hunk 3]
```diff
@@ -320,232 +320,136 @@ static int init_mix(struct comp_dev *dev,
...
-					      audio_stream_sample_bytes(&sink_c->stream));
-		buffer_stream_writeback(sink_c, sink_bytes);
-
-		comp_update_buffer_produce(sink_c, sink_bytes);
-		comp_update_buffer_consume(source_c, source_bytes);
+		sink_sample_bytes = audio_stream_sample_bytes(mod->output_buffers[0].data);
+		audio_stream_copy_from_linear(cd->buf_out, 0, mod->output_buffers[0].data, 0,
+					      sink_bytes / sink_sample_bytes);
+		mod->output_buffers[0].size = sink_bytes;
+		mod->input_buffers[0].consumed = sink_bytes;
```
Comment (abonislawski) [inline src/audio/up_down_mixer/up_down_mixer.c:441]: This is intentional?

comp_update_buffer_produce(sink_c, **sink_bytes**);
comp_update_buffer_consume(source_c, **source_bytes**);

->

mod->output_buffers[0].size = **sink_bytes**;
mod->input_buffers[0].consumed = **sink_bytes**;

Comment (ranj063) [reply src/audio/up_down_mixer/up_down_mixer.c]: no not at all. Thanks for catching this!

File: src/audio/up_down_mixer/up_down_mixer.c (hunk 2 above)
Comment (ranj063) [reply src/audio/up_down_mixer/up_down_mixer.c:434]: @kv2019i yes this is something that needs to cleand up in all modules. Will follow up in another PR

Comment (btian1) [reply src/audio/up_down_mixer/up_down_mixer.c:434]: I can do this, I created one jira to track this: https://jira.devtools.intel.com/browse/SOFC2-311

only one question, shall we do it right now, or wait for all component move to moduler adapter?

**Comment by lgirdwood** [comment]
@ranj063 still seeing a CI failure - expected ?

**Comment by ranj063** [comment]
> @ranj063 still seeing a CI failure - expected ?

no @lgirdwood. I'm still working with @abonislawski to figure out what's going on.

File: src/audio/up_down_mixer/up_down_mixer.c (hunk 2 above)
Comment (ranj063) [reply src/audio/up_down_mixer/up_down_mixer.c:434]: @btian1 we can start doing it now.

File: src/audio/up_down_mixer/up_down_mixer.c [hunk 4]
```diff
@@ -235,7 +235,7 @@ static up_down_mixer_routine select_mix_out_5_1(struct comp_dev *dev,
 			return upmix16bit_2_0_to_5_1;
 		case IPC4_CHANNEL_CONFIG_INVALID:
 		default:
-			comp_err(dev, "select_mix_out_mono(): invalid channel config.");
```
[the remark this answers is not in the document]
Comment (kv2019i) [reply src/audio/up_down_mixer/up_down_mixer.c:238]: Sure, just noting. Kind of distracts when reviewing

**Review by abonislawski — APPROVED** [review]
All good now

**Comment by ranj063** [comment]
> All good now

Thank you so much @abonislawski

File: src/audio/up_down_mixer/up_down_mixer.c (hunk 2 above)
Comment (btian1) [reply src/audio/up_down_mixer/up_down_mixer.c:436]: ok, let me submit a PR for this topic.

**Comment by lgirdwood** [comment]
Rerun CI, one of the DUTs has been having issues.

**Comment by lgirdwood** [comment]
SOFCI TEST

**Comment by kv2019i** [comment]
https://github.com/thesofproject/sof/issues/6864 his once in https://sof-ci.01.org/sofpr/PR6892/build3433/devicetest/index.html -> this is expected and should not block merge.

@ranj063 how about this one, this seems new to me (platform does wake up from suspend but aplay/record has error):
https://sof-ci.01.org/sofpr/PR6892/build3433/devicetest/index.html?model=ADLP_RVP_SDW_IPC4ZPH&testcase=check-suspend-resume-with-playback

**Comment by ranj063** [comment]
> #6864 his once in https://sof-ci.01.org/sofpr/PR6892/build3433/devicetest/index.html -> this is expected and should not block merge.
> 
> @ranj063 how about this one, this seems new to me (platform does wake up from suspend but aplay/record has error): https://sof-ci.01.org/sofpr/PR6892/build3433/devicetest/index.html?model=ADLP_RVP_SDW_IPC4ZPH&testcase=check-suspend-resume-with-playback

@kv2019i the updownmixer is not used in any of our tplgs. so if there's a failure it must have been seen with SOF main already

**Comment by kv2019i** [comment]
@ranj063 wrote:
> @kv2019i the updownmixer is not used in any of our tplgs. so if there's a failure it must have been seen with SOF main already

Ack, then let's proceed and merge.

**Pull request merged**